### ISSUE TITLE
Access Merkle Proof in LazyIMT

### ISSUE THREAD

**Issue by alrevuelta** [privacy-scaling-explorations/zk-kit#123]
Question rather than feature, but it might lead to a feature if feasible.

I'm using the [LazyIMT](https://github.com/privacy-scaling-explorations/zk-kit/blob/96ceeba751756d8abaa179ad5c52afecf25c3fc7/packages/imt.sol/contracts/internal/InternalLazyIMT.sol) and was wondering if it would be possible to **get a Merkle Proof directly from the contract**. The main benefit I see, is that any app would be able to get said proof to prove things, without having to locally fetch the whole tree and construct it locally. I'm using the `LazyIMT` because it seems to offer the closest I'm looking for, but open for other trees.

With a reduced example, given the following tree with 8 leafs, I would like to get a Merkle Proof for a given leaf, **provided by the contract**. Example:
* Proof that `C` belongs to the tree: [`D`, `HASH_AB`, `HASH_EFGH`]
* Proof that `A` belongs to the tree: [`B`, `HASH_CD`, `HASH_EFGH`]

```
          ROOT (Hash_ABCDEFGH)
          /                  \
         /                    \
 HASH_ABCD                   HASH_EFGH
  /      \                   /       \
 /        \                 /         \
HASH_AB   HASH_CD      HASH_EF    HASH_GH
 /   \     /   \       /   \      /   \
A     B   C     D     E     F    G     H

```

The `LazyIMT` seems to store in [`elements`](https://github.com/privacy-scaling-explorations/zk-kit/blob/96ceeba751756d8abaa179ad5c52afecf25c3fc7/packages/imt.sol/contracts/internal/InternalLazyIMT.sol#L10) the whole tree, both leafs at the bottom and intermediate levels. In theory, I could get the Merkle Proof of `C`, by accessing `elements` at different levels:
  * `D`: `_indexForElement(0, 3)`
  * `HASH_AB`: `_indexForElement(1, 0)`
  * `HASH_EFGH`: `_indexForElement(2, 1)`

The problem is that when the leaf is a [left element](https://github.com/privacy-scaling-explorations/zk-kit/blob/96ceeba751756d8abaa179ad5c52afecf25c3fc7/packages/imt.sol/contracts/internal/InternalLazyIMT.sol#L114), it's not hashed. So when the amount of leaves is odd, then upper hashes don't match, since they are not initialized yet. In other words, in the above example, one would expect `element` at `_indexForElement(3, 0)` to be the root, but this holds true only when the amount of leafs is even. Based on this, seems like I can't naively make `elements` public and use it for my purpose.

**Some questions:**
* Why [not hashing](https://github.com/privacy-scaling-explorations/zk-kit/blob/96ceeba751756d8abaa179ad5c52afecf25c3fc7/packages/imt.sol/contracts/internal/InternalLazyIMT.sol#L114) when having only a left element?
* How gas efficient is a `LazyIMT` when compared to others? As far as I understand `elements` stores both leaves and intermediate ones, hashed up to the root. This means that in my example (8 leafs) a total of 15 elements are stored (8,4,2,1) and updated. Can an insert trigger up to `MAX_DEPTH` hashes?

**tldr:** In order to not fall into the [x->y](https://xyproblem.info/), this is the problem I'm trying to solve. I would like to get the Merkle Proof of a given leaf, ideally with a view function and without worsening the gas consumption. Is this feasible? Which tree would you recommend me to use? Any idea on how to implement this?

Thanks.

**Comment by vimwitch** [comment]
Yes, this is possible to do. I'll answer your numbered questions first.

1. The "lazy" tree performs the minimum number of hashes necessary to insert elements in a tree. This means if there is only a left element the parent hash is not calculated until a corresponding right element exists, to avoid having an intermediate hash that will change in the future. This tree is designed for roots that are _infrequently_ accessed onchain. The specific case we had was a tree that was progressively built over a period of time with the root only accessed once at the end.
2. The lazy approach is ~25% cheaper over the lifetime of the tree. The approach is much more efficient for mostly empty/small trees and equally efficient for tree that are nearly full (assuming the tree root is infrequently accessed). If the root needs to be accessed often this approach is less efficient. The cost of calculating the root is (in the worst case) the same as the cost of inserting into a normal tree. You can see the original conversation with some rough benchmarks [here](https://github.com/Unirep/Unirep/pull/416#issuecomment-1515751236). There's a decent description of the tree [here](https://github.com/Unirep/Unirep/issues/411).

To answer the original question, yes this is possible. As you noticed some hashes don't exist in the tree, but the hash of the elements necessary to build the proof should always exist. To build a merkle proof the contract will need to iterate up the tree and at each level check if the intermediate root exists. If the root does not exist the intermediate root should be calculated and used when building higher level roots. This implementation doesn't include such a function because we didn't need it at the time. Building something like this shouldn't be too hard, it will be similar to the update and root functions. I'll look into it later today and see if i can write something simple.

**Comment by cedoor** [comment]
> Working on #162 addressing this feature.

Thanks @alrevuelta 👍🏽  Tracking this on the ZK-Kit board!
